### DISCUSSION TITLE
Proposal for adaptive regulation and calibration of target temperature

### DISCUSSION THREAD

**Discussion by maia** [jmcollin78/versatile_thermostat#208] [Ideas]
Here is my proposal for additions and adjustments of the current algorithm of "self-regulated target temperature". I would appreciate any kind of input and feedback, because the more people think about the logic, the better it will be once it is implemented:

---

### 1. Calibration offset
This value is the long term average of the difference between target temperature and room temperature. It is relevant for all heating methods using a temperature sensor close built into the heating device, such as a thermostatic radiator valve (TRV), because the difference between this internally measured temperature and the target temperature directly influences the valve opening in TRVs (and possibly the heat output in other heating systems, please update this proposal for the final documentation).

This calibration offset does not swing wildly, it will rather stay at the same value over many days, weeks or even months.

To calculate the calibration offset we should ignore a period of time after a change in target temperature, as usually it then will either be warmer or colder than the target temperature. One option is to ignore anything in the first 2 (?) hours after a change in target temperature (*Does someone have a better idea that’s simple to implement?*). 

To calculate the calibration offset we will feed a EMA algorithm for irregular time series with a very long half life (does 3 days sound okay?) with the delta of target_temperature and current_temperature (with the exclusion mentioned above) and the timestamp. This calibration offset should be persisted across reboots of Home Assistant as it takes many days to build up.

*Open question: should the short- and mid-term regulation values have an effect on the delta calculation?*

### 2. Reaction to external heat sources
Rooms can overheat for many hours in a row due to various reasons: strong sunshine, cooking, many people in a room and also an open door leading to a room with a higher temperature.

Ideally versatile_thermostat can anticipate sudden increases in room temperature and react to that. My current suggestion is to calculate a 60 minute slope of the measured (and EMA smoothed) temperature and use this to estimate the increase in room temperature over the next 10 minutes. So if the room temperature increased by 1.5-2.0°C over the past hour, we can assume that the temperature will rise another 0.3°C in the next hour. This value should be considered when calculating the adjusted target temperature.

*Open question: which effect does the thermal mass (heat capacity) of the building have on the duration of the slope calculation and prediction time frame? Would it make sense to have a „fuzzy“ configuration option regarding the thermal mass and heating system (e.g. „how long does it take to heat a room by 1°, is it rather 5 minutes, 15 minutes or 60 minutes?“), which will then influence various parameters? This could be used instead of manually asking users to define cycle lengths*

### 3. Influence of external temperature
There should be a user configurable option regarding the effect of the external temperature. If it’s very, very cold outside, does a user want the indoor temperature to be *lower* than usual (to save energy) or *higher* than usual (for comfort)?

### 4. Dealing with open windows
*Open question: depending on the thermal mass (heat capacity) of the building the room temperature can quickly revert once a window is closed. So what shall happen when a window is opened? And what, when it is closed? Should it influence the regulated target temperature?*

### 5. Short term regulation
*Open question: This is something the PI-algorithm is currently used for. I have zero experience with it, probably it’s best to just keep the strong/medium/light settings?*
*Open question: can/should we use the internal TRV temperature somehow? Because the TRV will keep its valve rather closed when the temperature measured by the TRV is above the target temperature. It does not know/care about the room temperature. So if we want to heat a room, the target temperature would need to be slightly *above* the temperature measured by the TRV itself. This could possibly mean that when using the "over climate" mode one will not use the current PI-algorithm but instead something that also adjusts the target temperature based on the currently measured TRV temperature?*

### Total adjustment calculation
I suggest to add all the adjustments listed above and then optionally limit the result with a range of e.g. -3°C to +3°C compared to the current target temperature.

---

What do you think? I will update this proposal based on your reactions and knowledge about different heating systems. Thanks!

**Comment by MaxWaldorf** [comment]
I would add a comment for Heat Pumps to have a lower target than the set temperature because their calibration isn't always great and can rapidly overheat a room...
Maybe the calculation should take this into account when a room is over the limit for a certain amount of time?

**Reply by maia** [reply]
Thanks for your feedback! The regulation should look at long term and short term differences between target temperature and actual temperature and adjust accordingly. It's not just heat pumps having such issues, it's also all TRVs that control their valve opening based on the temperature measured internally (next to the radiator), that's where the regulation should shine. So it's partly a technical question, partly a proper guidance of the users when configuring everything.

**Reply by MaxWaldorf** [reply]
Yeah but configuring a heat pump or TRV shouldn't require a master's degree and a NASA position :)

**Reply by maia** [reply]
Exactly. I still envision a configuration screen with less detail options and a more general question like "what's the latency of your heating system, what's the thermal mass of your building, how many minutes/hours does it take to heat the room by 1°?" and many details are calculated internally based on this information.

**Comment by xoferz** [comment]
One of the issues I have discovered that needs manual tinkering with Versatile Thermostat for me, is on/off time. I run an underfloor water heating and actuators take a couple of minutes to open or close. I have been tinkering with cycle duration and other parameters, but still sometimes when temperature is close to target the thermostat "trickles" so often that the actuators do not have time to actually open or close. I would love to enter time to open/close as a separate parameter. In my case, it is about 2 minutes. I tested at a friends house, and his actuators take 5 minutes to open or close.

**Comment by CodingGhost** [comment]
I also have the Problem with the wrong temperature on the TRV. I have solved it using an automation that calibrates the TRV in addition to VT, but ideally VT would do this by itself ofcourse. The Automation I used is from a blueprint that can be found here:
https://github.com/rotilho/home-assistant-blueprints/blob/main/trv_calibrator.yaml

**Reply by maia** [reply]
Thanks. Some TRVs expose this calibration offset to HA, unfortunately not all do (e.g. Eve Thermo). It would be great to have some logic that will adjust accordingly in Versatile Thermostat.

**Comment by JoryHogeveen** [comment]
Would also be nice to have an overshoot limit (opt-in) setting of some sorts. Once the target temperature has been reached, enforce the target climate entity to shut down (lower it's target temperature).

**Reply by CodingGhost** [reply]
Well thats up to the PI controller, if it overshoots too much, lower your Proportional(P) parameter.

**Reply by JoryHogeveen** [reply]
Thanks for the tip @CodingGhost I've lowered the Kp param in the DIYLESS thermostat to 10 (instead of 30)
